### PR TITLE
Run API Proctor tests in build-ship-test

### PR DESCRIPTION
**Pull request by jonathansharman**
### Documentation

- [Asana task](https://app.asana.com/0/1202563967140527/1202561384547922/f)
- [Back-End Patterns & Practices / Continuous Integration](https://bookstack.niche.team/books/back-end-patterns-practices/page/continuous-integration) - BookStack article explaining how prepare a service to use this workflow
- [BE Service CI Status sheet](https://docs.google.com/spreadsheets/d/18JOH27RaSS3KXUxSqEmTYZzPjYRig-DNTTJ1s4kqzrc/edit#gid=0) - lists the current status and next steps for each (Go) BE repo
- Adapted from https://github.com/nicheinc/entity/pull/221

### Description

This PR extends the `build-test-ship` workflow to run [API Proctor](https://github.com/nicheinc/api-proctor) tests. I've left numerous comments on the workflow itself, but I'll also call out some caveats and future work here. Reviewers, don't hesitate to message me or post in `#eso-delta` if you'd like to talk through anything in here!

#### Submodule Checkout

At the moment, this checks out all the `development` submodules so that all the files specified in the `docker-compose.yml` are present (see [this slack thread](https://nicheinc.slack.com/archives/C026TLM8HSL/p1662047480315389)). This takes some time (around 1.5 minutes). Cochran had thoughts in the thread on how to possibly avoid this in the future by eliminating `docker.env` files. There is an [Asana task](https://app.asana.com/0/1201453510342377/1202915568704245/f) for that work, which also describes a related problem where the workflow does not see the current branch's updates to the `docker.env`.

#### GitHub Artifacts

The workflow uses [artifacts](https://docs.github.com/en/actions/using-workflows/storing-workflow-data-as-artifacts) to store service images built from the current branch of the current repo (identified via the `services` workflow input) so they can be pulled during the `run-api-proctor-tests` job. We only need these images for the duration of the run, but surprisingly there is no first-party action for deleting artifacts. I've set the retention period for the docker image artifacts to 1 day, which I think will clear them soon enough that we won't run into storage limits. However, if we decide we'd like to clear the artifacts immediately, we could use [this API](https://docs.github.com/en/rest/actions/artifacts#delete-an-artifact) or possibly one of the third-party `delete-artifacts` actions floating around.

#### Branches of Dependencies

Except for services from the current repo, the workflow always uses the `latest` service images from ECR, so it has no way to use images from branches of dependencies (e.g. in PRs that depend on PRs in other repos). This could be a little inconvenient during development because tests might fail in CI until the dependency has been deployed. However, by the time the consuming branch is merged, the dependency should be deployed, so we will still have useful checks by the time the dependent PR is merged. This is also only an issue if behavior invoked by _existing_ `api-proctor` tests depends on the updated behavior from the other branch, which is probably not very common. At least for MVP, I don't think this workflow should support branches of dependencies due to the complexity involved.

#### Service Launch/API Proctor Timing

Some services have dependencies, such as Elasticsearch, that can take a while to fully start. Running the API Proctor tests immediately after launching the services with `docker-compose` can result in test failures because, by default, `docker-compose up -d` doesn't wait for services to actually be ready/healthy before returning. Currently, the workflow works around this by [sleeping for one minute](https://github.com/nicheinc/actions/blob/e5be49c5d565184eed55e1357a8d458b402ad48f/.github/workflows/build-test-ship.yaml#L380) between launching the services and running the API Proctor tests, which is usually overkill. We have a [backlog task](https://app.asana.com/0/1201010009456007/1203332769997382/f) that describes the issue (and what we've tried so far) in more detail.

### Testing Considerations

The pilot service branches for this workflow are https://github.com/nicheinc/entity/pull/241, https://github.com/nicheinc/fact/pull/324, and https://github.com/nicheinc/search/pull/238, but we have open branches in all (Go) service repos. The [BE Service CI Status sheet](https://docs.google.com/spreadsheets/d/18JOH27RaSS3KXUxSqEmTYZzPjYRig-DNTTJ1s4kqzrc/edit#gid=0) shows the current status and next steps for each repo.

We should make sure I haven't introduced any regressions during merges, especially in the deployment steps, which are difficult/impossible to test during workflow development.

### Deployment 

The updated workflow from this PR has already been copied into a new repo, https://github.com/nicheinc/actions-go-ci, as described in [this subtask](https://app.asana.com/0/0/1203340028733045/f). Once all CRs are in, we should make any necessary changes over there and close this PR and delete its branch.

We will roll this out to each service repo one-by-one by pointing each service's `ci.yaml` to `nicheinc/actions-go-ci/.github/workflows/action.yaml@v0.1.0`.

### Versioning

This will be version 0.1.0 of a new repo.

### CODE REVIEW

File: .github/workflows/build-test-ship.yaml [hunk 1]
```diff
@@ -128,6 +146,143 @@ jobs:
...
+
+      - name: View kafka logs, for debugging
+        shell: bash
+        run: |
+          docker-compose ps | grep kafka
+          docker-compose logs kafka
+
+      - name: View entity logs, for debugging
+        shell: bash
+        run: docker-compose logs entity
```
Comment (angelowilliams) [inline .github/workflows/build-test-ship.yaml]: Since we're implementing this workflow in `fact` now too, we should generalize this. I think this would work (unless you can't have variables in step names?):
```
      - name: View ${{ matrix.service }} logs, for debugging
        shell: bash
        run: docker-compose logs ${{ matrix.service }}
```

Comment (jonathansharman) [reply .github/workflows/build-test-ship.yaml]: Absolutely - that step is just for debugging and will be removed entirely before I put this up for final CR. There's kind of a lot of jank hanging around in this branch right now. (Pay no attention to the man behind the curtain! 🙈)

File: .github/workflows/build-test-ship.yaml [hunk 2]
```diff
@@ -25,7 +29,7 @@ on:
         required: true
       ECR_SECRET_ACCESS_KEY:
         required: true
-      GHA_DEV_KEY:
+      NPM_WRITE_TOKEN:
```
Comment (jonathansharman) [inline .github/workflows/build-test-ship.yaml:36]: `NPM_WRITE_TOKEN` is a [personal access token](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token) for the `deployatniche` user, which we need in order to have permission to check out submodules from the `development` repo in the `run-api-proctor-tests` job below. See [this slack thread](https://nicheinc.slack.com/archives/C0FTZM5H7/p1662064211591039).

File: .github/workflows/build-test-ship.yaml [hunk 3]
```diff
@@ -76,11 +80,10 @@ jobs:
...
           go vet ./...
 
-  # Build the docker images on PR or push to dev/master, ship on release
-  build-and-ship-image:
-    name: "Build ${{ matrix.service }} image"
+  # Build images for api-proctor and, on release, ship final images to ECR.
+  build-and-ship-images:
+    name: Build ${{ matrix.service }} images
     runs-on: ubuntu-latest
-    if: ${{ !github.event.pull_request.draft }}
```
Comment (jonathansharman) [inline .github/workflows/build-test-ship.yaml:83]: I removed this guard so that we can run API Proctor tests on draft PRs, which I think is worthwhile. We can tweak the job conditions if we find we're running API Proctor more often than is useful.

File: .github/workflows/build-test-ship.yaml [hunk 4]
```diff
@@ -110,10 +113,10 @@ jobs:
...
             fi
           done
-          echo "::set-output name=appversion::${APPVERSION}"
-          echo "::set-output name=chartversion::${CHARTVERSION}"
-          echo "::set-output name=appchange::${appchange}"
-          echo "::set-output name=chartchange::${chartchange}"
+          echo "appversion=${APPVERSION}" >> $GITHUB_OUTPUT
+          echo "chartversion=${CHARTVERSION}" >> $GITHUB_OUTPUT
+          echo "appchange=${appchange}" >> $GITHUB_OUTPUT
+          echo "chartchange=${chartchange}" >> $GITHUB_OUTPUT
```
Comment (jonathansharman) [inline .github/workflows/build-test-ship.yaml:123]: The `set-output` command [is deprecated](https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/).

File: .github/workflows/build-test-ship.yaml [hunk 5]
```diff
@@ -131,21 +134,25 @@ jobs:
...
 
-      # On push/PR, ensure service images can be built on target platforms.
-      - name: Build final images
+      # Build and upload image artifact for use in api-proctor tests.
+      - name: Build image for api-proctor tests
         uses: docker/build-push-action@v2
-        if: |
-          github.event_name == 'pull_request' ||
-          (github.event_name == 'push' &&
-           steps.check-version-change.outputs.appchange == 'false')
```
Comment (jonathansharman) [inline .github/workflows/build-test-ship.yaml]: I've moved test builds (builds that just verify the service images can be built on the target platforms, without using the resulting images for anything) to a separate `build-for-target-platforms` job so they can run concurrently with the `build-and-ship-images` and `run-api-proctor-tests` jobs.

File: .github/workflows/build-test-ship.yaml [hunk 6]
```diff
@@ -131,21 +134,25 @@ jobs:
...
+          outputs: type=docker,dest=${{ matrix.service }}.tar
+          tags: ${{ matrix.service }}:ci
 
-      # On release, build service images and push to ECR.
+      - name: Upload docker image artifact
+        uses: actions/upload-artifact@v3
+        with:
+          name: ${{ matrix.service }}
+          path: ${{ matrix.service }}.tar
+          retention-days: 1
```
Comment (jonathansharman) [inline .github/workflows/build-test-ship.yaml]: Since jobs run on different machines (unless you're using custom runners), we can't trivially access files produced in one job from another. We have to upload the files in one job and download them in another, which we can do using GitHub artifacts, uploading TAR files containing images tagged with `:ci`. I've set the retention period to the lowest possible (1 day). See the PR writeup for caveats.

Ideally, we'd just build the services images in the `run-api-proctor-tests` job itself, to avoid the need for this inter-job communication. However, I'm not sure that's feasible because of the way we're using the `matrix` strategy to build all services. If we used a `matrix` strategy in `run-api-proctor-tests`, then instead of running the tests once with all _N_ of the branch's services running, we'd end up running the tests _N_ times - and each of those runs would only be using the branch's version of _one_ service at a time.

File: .github/workflows/build-test-ship.yaml [hunk 7]
```diff
@@ -183,34 +190,251 @@ jobs:
...
+    name: Run API Proctor tests
+    runs-on: ubuntu-latest
+    needs: build-and-ship-images
+    timeout-minutes: ${{ inputs.api-proctor-timeout-minutes }}
+    env:
+      PGADMIN_PORT: 5000
+      POSTGRES_USER: postgres
+      POSTGRES_PASSWORD: postgres
+      POSTGRES_SERVER: postgres
+      POSTGRES_PORT: 5432
```
Comment (jonathansharman) [inline .github/workflows/build-test-ship.yaml:238]: Maybe it would be better to put these in a `.env.ci` file in the `development` repo, or something like that?

File: .github/workflows/build-test-ship.yaml [hunk 8]
```diff
@@ -183,34 +190,251 @@ jobs:
...
+        run: |
+          tools=`yq -o=json '.services
+            | with_entries(select(.value.profiles.[] | contains("tools")))
+            | to_entries
+            | map(.key)' docker-compose.yml`
+          all_services=`yq -o=json '.services | keys' docker-compose.yml`
+          non_tools=`jq --null-input "${all_services} - ${tools}"`
+          input_services='${{ inputs.services }}'
+          services=`jq --null-input "${input_services} - (${input_services} - ${non_tools})"`
+          echo "services=`jq -r 'join(" ")' <<< ${services}`" >> $GITHUB_OUTPUT
```
Comment (jonathansharman) [inline .github/workflows/build-test-ship.yaml:265]: Some repos include CLI tools in the `services` input because those tools need to be built and shipped to ECR. However, we don't want to run these tools as part of this job (and some of them aren't even in the `docker-compose.yml`), so we need to filter out anything that's not a non-"tool" service. I've done that by doing some set operations to get the intersection of the `services` input and the `docker-compose` services that aren't part of the `tools` profile.

This is script is functional but horrible because I'm bad at `yq` and `jq`. Suggestions for simplifying this step appreciated!

File: .github/workflows/build-test-ship.yaml [hunk 9]
```diff
@@ -183,34 +190,251 @@ jobs:
...
+      - name: Create docker-compose override and AWS credentials files
+        env:
+          ENCODED_CREDENTIALS: ${{ secrets.DEV_AWS_CREDENTIALS }}
+        shell: bash
+        run: |
+          yq -n ".version = \"3.7\"" > docker-compose.override.yaml
+          for service in ${{ steps.get-services.outputs.services }}; do
+            yq -i ".services.${service}.image = \"${service}:ci\"" docker-compose.override.yaml
+          done
+          yq eval-all --inplace 'select(fileIndex == 0) * select(fileIndex == 1)' docker-compose.override.yaml ${{ github.workspace }}/docker-compose.ci-override.yml
```
Comment (jonathansharman) [inline .github/workflows/build-test-ship.yaml:300]: These lines serve two purposes:
1. Overriding the image for all services in this repo to use the `:ci`-tagged image we built in an earlier job, instead of the latest images in ECR
2. Merging [`docker-compose.ci-override.yml`](https://github.com/nicheinc/development/blob/master/docker-compose.ci-override.yml) into `docker-compose.override.yaml`* to pick up any special changes this repo requires, such as using a local `elasticsearch` container instead of `dev-elasticsearch`

*Apologies for the mixed file extensions. Eventually we should standardize the whole `development` repo around `.yaml`, for consistency.

File: .github/workflows/build-test-ship.yaml [hunk 10]
```diff
@@ -183,34 +190,251 @@ jobs:
...
+          echo $ENCODED_CREDENTIALS | base64 -di > vault/aws_credentials
+
+      # TODO: This is a hack to place empty secret files in the vault directory
+      # (to satisfy docker-compose.yml requirements) without actually retrieving
+      # them from vault.
+      - name: Retrieve vault secrets
+        shell: bash
+        run: |
+          touch vault/website.secrets.env
+          touch vault/landingpageadmin.secrets.env
```
Comment (jonathansharman) [inline .github/workflows/build-test-ship.yaml:312]: Without this step, `docker-compose` chokes, even though we don't actually invoke the services that need these files. 😕

File: .github/workflows/build-test-ship.yaml [hunk 11]
```diff
@@ -183,34 +190,251 @@ jobs:
...
+            deps[${next}]=1
+            next_deps=$(yq ".services.${next}.depends_on[]" docker-compose.yml)
+            for next_dep in ${next_deps}; do
+              if [[ ! ${deps[${next_dep}]} ]]; then
+                work_set[${next_dep}]=1
+              fi
+            done
+          done
+          echo Dependencies: ${!deps[@]}
+          echo "deps=${!deps[@]}" >> $GITHUB_OUTPUT
```
Comment (jonathansharman) [inline .github/workflows/build-test-ship.yaml:340]: In the next step, we need to create databases and roles for all the applicable services we're going to run during the API Proctor tests. That step is time-intensive, so we don't want to do it for more services than we need to. We only need the services from this repo (specified by the `services` input) and all of their transitive dependencies.

As far as I know, there's no built-in command for extracting a set of transitive dependencies from a `docker-compose.yaml`, so I wrote this graph-traversal bash script (which was fun, but also 🤮). All ears if someone knows a better way to do this!

File: .github/workflows/build-test-ship.yaml [hunk 12]
```diff
@@ -183,34 +190,251 @@ jobs:
...
+
+      - name: Start postgres and create service databases and roles
+        shell: bash
+        run: |
+          docker-compose up -d postgres
+          timeout 1m bash -c 'until docker-compose run postgres pg_isready -d postgres://${POSTGRES_USER}:${POSTGRES_PASSWORD}@postgres; do sleep 1s; done'
```
Comment (jonathansharman) [inline .github/workflows/build-test-ship.yaml:346]: This should be exactly the kind of thing that [healthchecks](https://docs.docker.com/compose/compose-file/compose-file-v3/#healthcheck) and the [`--wait` option](https://docs.docker.com/engine/reference/commandline/compose_up/#options) of `docker compose up` are for. However, when I tried that ([`actions` PR](https://github.com/nicheinc/actions/pull/8/files), [`development` branch](https://github.com/nicheinc/development/compare/master...x/pg-healthcheck)), `docker compose` appeared to [give up immediately](https://nicheinc.slack.com/archives/C026TLM8HSL/p1666027836893279?thread_ts=1666016597.341929&cid=C026TLM8HSL) upon seeing that `postgres` was unhealthy, rather than retrying until it was healthy. (Worse, it failed inconsistently.) Let me know if I was just just holding it wrong.

File: .github/workflows/build-test-ship.yaml [hunk 13]
```diff
@@ -183,34 +190,251 @@ jobs:
...
+          declare -A dbs
+          conn_regex='postgres://(\w*):(\w*)@\$\{POSTGRES_SERVER\}/(\w*)'
```
Comment (jonathansharman) [inline .github/workflows/build-test-ship.yaml:348]: This step uses the existence of a `POSTGRES_CONNECTION` variable in the `docker-compose.yml` to infer that a service has a database - and if so to create the database and execute its `roles.sql` and `extensions.sql` scripts.

File: .github/workflows/build-test-ship.yaml [hunk 14]
```diff
@@ -183,34 +190,251 @@ jobs:
...
+        run: |
+          docker run --rm --name api-proctor --network host \
+            -e READ_POSTGRES_CONNECTION=postgres://${POSTGRES_USER}:${POSTGRES_PASSWORD}@localhost:${POSTGRES_PORT} \
+            -e WRITE_POSTGRES_CONNECTION=postgres://${POSTGRES_USER}:${POSTGRES_PASSWORD}@localhost:${POSTGRES_PORT} \
+            -e NODE_ENV=ci \
+            $ECR_REGISTRY/api-proctor:latest \
+            tests/${{ github.event.repository.name }}/ --testPathIgnorePatterns=/src/ --passWithNoTests
+
+      - name: Docker Compose logs
+        if: success() || failure()
```
Comment (jonathansharman) [inline .github/workflows/build-test-ship.yaml:399]: We always want to show logs - even if the tests fail - to give us more debugging info.

File: .github/workflows/build-test-ship.yaml [hunk 15]
```diff
@@ -183,34 +190,247 @@ jobs:
...
             ${{ env.ECR_REGISTRY }}/${{ env.ECR_REPOSITORY }}:latest
 
+  build-for-target-platforms:
+    name: Verify ${{ matrix.service }} images can be built on target platforms
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        service: ${{ fromJson(inputs.services) }}
+    if: github.event_name == 'pull_request' && !github.event.pull_request.draft
```
Comment (jonathansharman) [inline .github/workflows/build-test-ship.yaml:207]: I've restricted this job to non-draft PRs because it takes a long time and in my opinion doesn't provide much signal but is still worth running at least once before release.

File: .github/workflows/build-test-ship.yaml (hunk 2 above)
Comment (evan-kochuk-niche) [reply .github/workflows/build-test-ship.yaml:36]: Was `GHA_DEV_KEY` Defunct? I don't see any other changes to it, and it's not in the current version of the file. Not a major factor either way, just curious.

File: .github/workflows/build-test-ship.yaml [hunk 16]
```diff
@@ -131,21 +134,25 @@ jobs:
...
           context: .
           file: ./cmd/${{ matrix.service }}/Dockerfile
           cache-from: type=gha
           cache-to: type=gha,mode=max
-          platforms: linux/amd64,linux/arm64
+          outputs: type=docker,dest=${{ matrix.service }}.tar
+          tags: ${{ matrix.service }}:ci
 
-      # On release, build service images and push to ECR.
+      - name: Upload docker image artifact
```
Comment (evan-kochuk-niche) [inline .github/workflows/build-test-ship.yaml:154]:
```suggestion
      # Since jobs run on different machines (unless you're using custom runners), we can't trivially access files produced in one job from another.
      # We have to upload the files in one job and download them in another, which we can do via GitHub
      # Artifacts, which we upload as tagged .TAR files
      - name: Upload docker image artifact
```

**Review by evan-kochuk-niche — COMMENTED** [review]
Though there's a couple of self-described "Hacks" throughout the PR, none of the problems surrounding those are trivial, nor do I have an immediate insight in how to moot them. Given the relative magnitude of what we're attempting to implement here, I think the more eyes, the better, so while I haven't found anything that needs immediate redress (excepting a single place where I think one of Sharman's clarifying PR comments ought to be included in the file itself), I'm going to hold out on explicit approval until more eyes have considered this.

File: .github/workflows/build-test-ship.yaml (hunk 16 above)
Comment (jonathansharman) [reply .github/workflows/build-test-ship.yaml:154]: > clarifying PR comments ought to be included in the file itself

Good idea! Reworded and added in ec4c8fdda385d0ec963b0f24388a9f4e96c4d0e5

File: .github/workflows/build-test-ship.yaml (hunk 2 above)
Comment (jonathansharman) [reply .github/workflows/build-test-ship.yaml:36]: In more detail than is necessary... That was a relic from when I first factored out `build-test-ship` into a single reusable workflow. I actually started out with the `feature/api-proctor` branch and then cut out the API Proctor stuff. When I did that, I accidentally left that secret in the requirements, even though we weren't using it yet, and I just didn't bother fixing it because it didn't hurt, and I thought we'd need it later (this PR) anyway, to clone the `development` environment.

But it turned out that `GHA_DEV_KEY` wasn't powerful enough: it doesn't have read access to all the submodules, which we ended up needing, so I swapped it out for `NPM_WRITE_TOKEN`.

**tl;dr: `GHA_DEV_KEY` was never actually used.**

File: .github/workflows/build-test-ship.yaml [hunk 17]
```diff
@@ -183,34 +197,248 @@ jobs:
...
+          platforms: linux/amd64,linux/arm64
+
+  run-api-proctor-tests:
+    name: Run API Proctor tests
+    if: inputs.run-api-proctor-tests
+    runs-on: ubuntu-latest
+    needs: build-and-ship-images
+    timeout-minutes: ${{ inputs.api-proctor-timeout-minutes }}
+    env:
+      PGADMIN_PORT: 5000
```
Comment (dzbee) [inline .github/workflows/build-test-ship.yaml:234]: Do we need pgAdmin? I understand we might spin it up as part of the development environment but I don't expect we'd need to talk to it during CI. If we could not even spin it up that would be great. I don't know if we're spinning up the whole dev environment or just BE or relying on the listed dependencies in `docker-compose.yml`; it wasn't clear to me looking at the actions that build the images.

File: .github/workflows/build-test-ship.yaml (hunk 7 above)
Comment (dzbee) [reply .github/workflows/build-test-ship.yaml:238]: Is the idea that the source of truth would then live in `development`, where we otherwise already control what ports things run on? I think it's fine to have here, personally, but don't have any strong opinion if you want it to be otherwise.

File: .github/workflows/build-test-ship.yaml [hunk 18]
```diff
@@ -183,34 +197,248 @@ jobs:
...
+        shell: bash
+        run: |
+          tools=`yq -o=json '.services
+            | with_entries(select(.value.profiles.[] | contains("tools")))
+            | to_entries
+            | map(.key)' docker-compose.yml`
+          all_services=`yq -o=json '.services | keys' docker-compose.yml`
+          non_tools=`jq --null-input "${all_services} - ${tools}"`
+          input_services='${{ inputs.services }}'
+          services=`jq --null-input "${input_services} - (${input_services} - ${non_tools})"`
```
Comment (dzbee) [inline .github/workflows/build-test-ship.yaml:264]: I'm a little confused about why we've written this expression this way. Assuming `input_services` is a subset of `all_services`, then I think this expression must always be equivalent to `input_services - tools`. If I'm interpreting the expression and intent correctly, I think `input_services - tools` would be much clearer (admittedly I'm a little confused by the double-negative in your Github comment below).

Alternatively, if `input_services` was a single string (rather than an array of services) I wrote a little expression that would just get the service and its dependencies:
```yq -o=json '.services | with_entries(select(.key | test($input_service))) | (.$input_service.depends_on) + [$input_service]' docker-compose.yml```
There's maybe some issues with that as-is (in particular I'm not sure how to get `$input_service` to evaluate for use as a key), and it would definitely need adaptation for plural input services, but I wanted to take a shot in the dark at something that might be more efficient.

Comment (dzbee) [reply .github/workflows/build-test-ship.yaml:264]: D'oh, of course we'll need the dependencies of the dependencies, and I see we traverse the graph later.

Also after more reading, I think I now understand that the issue is that service repos include tools, and we want to exclude tools, so `input_services` itself includes tools, in which case my new expression is not useful. However, I still think `input_services - tools` might be sufficient for what we have going on here at the moment, and more clear.

**Review by dzbee — APPROVED** [review]
LGTM, though there's a lot in here I'm unfamiliar with -- in general it looks like you're asking all the right questions for improving this moving forward, but I don't have any suggestions for now cause I don't know any better ways to do things. That said, nothing jumps out at me as being obviously wrong.

File: .github/workflows/build-test-ship.yaml (hunk 17 above)
Comment (jonathansharman) [reply .github/workflows/build-test-ship.yaml:234]: Good question! I tried taking it out in f7c2d7148fb11fa0aee722f31b6adbf8594a7152. It seems that even though we're not actually running `pgadmin4`, we have to give `PGADMIN_PORT` some value or else the [Compose file is malformed](https://github.com/nicheinc/entity/actions/runs/3534463670/jobs/5994927475#step:14:51). (I reverted the change in 7e98ce7d19c970d6404951e1950a87a7373e3eb0.)

File: .github/workflows/build-test-ship.yaml (hunk 18 above)
Comment (jonathansharman) [reply .github/workflows/build-test-ship.yaml:264]: **tl;dr: I don't think we can just do `input_services - tools` because not every "tool" in `inputs.services` is in `tools`.**

Sorry this algorithm and my explanation are confusing. 🙈 There are a few different sets of executables we're dealing with here:
- The `services` input the current repo passes to `build-test-ship`, which includes all executables that need to be built and shipped to ECR, including:
  - (A) Actual services (microservices, Kafka consumers, etc.)
  - (B) Non-service command-line tools
- Compose file services, including those that are:
  - (C) Not part of the `tools` profile
  - (D) Part of the `tools` profile

We're given A ∪ B (via `inputs.services`), C ∪ D (via `yq`), and D (via `yq`), and what we really want is A: just the services that (along with transitive dependencies) need to be running before we start the API Proctor tests. We also know that:
- A and B are disjoint (by definition)
- C and D are disjoint (by defintion)
- B and C are disjoint (since any tools that _are_ included are in fact marked with the `tools` profile)
- A ⊆ C (empirically, thankfully)

You'd think B ⊆ D, but unfortunately some executables are missing from the Compose file entirely. For instance, `lead` [passes `lead-scheduler`](https://github.com/nicheinc/lead/blob/801e256819ff0931a104c3c9caa4d0d55c796ad7/.github/workflows/ci.yaml#L32), which is absent from the [`docker-compose.yml`](https://github.com/nicheinc/development/blob/master/docker-compose.yml) in `development`. (Maybe we should add it there? But that's not the state of the world today, and there are about five other repos in a similar situation.)

So instead I think we have to do:
1. A = (A ∪ B) ∩ C, since A ⊆ C and B/C disjoint
2. C = (C ∪ D) - D, since they're disjoint
3. A = (A ∪ B) ∩ ((C ∪ D) - D), by (1) and (2)
4. A = A ∪ B - (A ∪ B - ((C ∪ D) - D)), by [math](https://proofwiki.org/wiki/Set_Difference_with_Set_Difference#:~:text=3%20Proof%202-,Theorem,T%E2%88%96(T%E2%88%96S))

That last step is necessary because there's no intersection operation built into `jq`.

File: .github/workflows/build-test-ship.yaml [hunk 19]
```diff
@@ -183,34 +197,248 @@ jobs:
...
+          done
+
+      - name: Launch services with dependencies via docker-compose
+        env:
+          REGISTRY: ${{ steps.login-ecr.outputs.registry }}
+        shell: bash
+        run: |
+          docker-compose pull --include-deps ${{ steps.get-services.outputs.services }}
+          docker-compose up -d ${{ steps.get-services.outputs.services }}
+          sleep 1m
```
Comment (bclarkx2) [inline .github/workflows/build-test-ship.yaml:384]: Small thought: ought this sleep timeout at least be configurable as an input to the workflow? It might be the kind of thing that is pretty reasonable to want to tweak on a per-service basis, depending on what the dependencies of each service are.

File: .github/workflows/build-test-ship.yaml [hunk 20]
```diff
@@ -183,34 +197,248 @@ jobs:
...
+      - name: Run API Proctor tests
+        env:
+          ECR_REGISTRY: ${{ steps.login-ecr.outputs.registry }}
+        shell: bash
+        run: |
+          docker run --rm --name api-proctor --network host \
+            -e READ_POSTGRES_CONNECTION=postgres://${POSTGRES_USER}:${POSTGRES_PASSWORD}@localhost:${POSTGRES_PORT} \
+            -e WRITE_POSTGRES_CONNECTION=postgres://${POSTGRES_USER}:${POSTGRES_PASSWORD}@localhost:${POSTGRES_PORT} \
+            -e NODE_ENV=ci \
+            $ECR_REGISTRY/api-proctor:latest \
```
Comment (bclarkx2) [inline .github/workflows/build-test-ship.yaml:395]: I think we are likely to have branches from time to time that will _not_ pass all of the `api-proctor:latest` tests and instead have a custom `api-proctor` branch tied to them that fixes the tests that the PR is breaking.

(This seems like it shouldn't happen because of backwards compatibility guarantees... but our `api-proctor` tests also do a lot of data seeding that is entirely liable to being broken by changes to database schemas).

I think this may put us in a place where the CI job on a PR fails even though the PR is "OK" given that it is bundled with an `api-proctor` branch that fixes any tests that may have been broken.

I think that's probably an acceptable state to be in for now, given that probably most PRs won't be in that situation. 

If we did want to provide a solution for this... I wonder if we could add a step that:
1. Checks for the existence of an `api-proctor` branch named something like `${{ github.event.repository.name }}/${GITHUB_REF##*/}` (e.g. using `gh` or something, if GHA doesn't have tools for this).
2. If found, checks out that `api-proctor` branch and runs it.
3. Otherwise, pulls `api-proctor:latest` from ECR and runs it.

If we are aiming to require CI to pass as a condition of merging a PR, then figuring out some solution here is probably going to be important, since I don't think expecting that `api-proctor` tests will never break will be an attainable goal. If not, I think we can probs just let this slide at least for now!

**Review by bclarkx2 — COMMENTED** [review]
This looks great! Just leaving a couple questions that arose for me during BENG.

File: .github/workflows/build-test-ship.yaml (hunk 19 above)
Comment (jonathansharman) [reply .github/workflows/build-test-ship.yaml:384]: Great thought! In fact, we probably want to default that to `0s` so we don't waste time in the repos that don't need this buffer - which I imagine will be most of them.

File: .github/workflows/build-test-ship.yaml (hunk 20 above)
Comment (jonathansharman) [reply .github/workflows/build-test-ship.yaml:395]: Excellent points! If I'm understanding your proposed solution, I think that would rely on naming the `api-proctor` branch the same as the service branch? That sounds workable, and I can't immediately think of a better way to associate the two. We could use a manual override, but then we'd have to remember to swap it back after the `api-proctor` branch is merged, which sounds error-prone.

File: .github/workflows/build-test-ship.yaml (hunk 19 above)
Comment (bclarkx2) [reply .github/workflows/build-test-ship.yaml:384]: Oh great point, yeah 0s seems like a good default to be tweaked upwards as needed!

File: .github/workflows/build-test-ship.yaml (hunk 20 above)
Comment (bclarkx2) [reply .github/workflows/build-test-ship.yaml:395]: Yeah, that's about the shape of my proposed hack :sweat_smile: This is probably something that BE and SEIT would need to hash out jointly, I'd imagine. Is there still a regular CI meeting going on that this could be discussed in? Regardless, I think it would be just fine to move forward with this PR and figure out this "api-proctor branch override" issue at a later date, since it seems like we are pretty far away from enforcing CI passing as a blocker to merging to `dev`.

> We could use a manual override, but then we'd have to remember to swap it back after the api-proctor branch is merged, which sounds error-prone.

Totally agreed -- anything that requires checking something into the branch that needs to be reverted on merge to `dev` seems worthwhile to avoid to me.

File: .github/workflows/build-test-ship.yaml (hunk 19 above)
Comment (jonathansharman) [reply .github/workflows/build-test-ship.yaml:384]: PR created in the new repo (just FYI - not implying you have to review it 😄): https://github.com/nicheinc/actions-go-ci/pull/1

File: .github/workflows/build-test-ship.yaml (hunk 18 above)
Comment (dzbee) [reply .github/workflows/build-test-ship.yaml:264]: Thank you for the explanation! It makes sense to me now.

File: .github/workflows/build-test-ship.yaml (hunk 20 above)
Comment (dzbee) [reply .github/workflows/build-test-ship.yaml:395]: This has maybe been discussed elsewhere but is there a compelling reason that API proctor test cases can't live in the service repos they test? I would assume this is the usual pattern for CI to avoid needing to sync branches across repos for test-breaking changes.

Comment (jonathansharman) [reply .github/workflows/build-test-ship.yaml:395]: I'm very interested in Dizzy's suggestion. 👀 For now I've added a [backlog task](https://app.asana.com/0/1201010009456007/1203467942453227/f) to track the problem and possible solutions. Thanks again for bringing this up. 👍

Comment (bclarkx2) [reply .github/workflows/build-test-ship.yaml:395]: :d'oh: Great point @dzbee ! It worked for playwright, so surely it can work for `api-proctor` as well :P

**Comment by jonathansharman** [comment]
Closing this in favor of https://github.com/nicheinc/actions-go-ci. If there are any other comments on this work, please open a ticket there or message ESO Delta!
